File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    # 1은 이동이 가능
    lx = len(maps)
    ly = len(maps[0])
    
    # 동서남북
    dx = [0, 0, 1, -1]
    dy = [1, -1, 0, 0]
    
    def bfs(maps, x, y):
        queue = deque()
        queue.append((x, y))
        
        while queue:
            x, y = queue.popleft()
            
            for i in range(4):
                # 4방향으로 모두 이동해본다.
                nx = x + dx[i]
                ny = y + dy[i]
                
                if nx < 0 or nx >= lx or ny < 0 or ny >= ly:
                    continue
                if maps[nx][ny] == 0:
                    continue
                if maps[nx][ny] == 1:
                    maps[nx][ny] = maps[x][y] + 1
                    queue.append((nx, ny))  
                    
    bfs(maps, 0, 0)

    if maps[lx-1][ly-1] == 1:
        return -1
    else:
        return maps[lx-1][ly-1]

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    
    class Node{
        int x;
        int y;
        int cnt;
        
        public Node(int x, int y, int cnt){
            this.x = x;
            this.y = y;
            this.cnt = cnt;
        }
    }
    
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    
    public int solution(int[][] maps) {
        
        int N = maps.length;
        int M = maps[0].length;
        
        boolean[][] visited = new boolean[N][M];
        Deque<Node> q = new ArrayDeque<>();
        int x = 0;
        int y = 0;
        
        visited[x][y] = true;
        q.offer(new Node(x, y, 1));
        
        while(!q.isEmpty()){
            Node cur = q.poll();
            
            if((cur.x == N-1) && (cur.y == M-1)){
                // 도착
                return cur.cnt;
            }
            
            for(int i=0; i<4; i++){
                int tx = cur.x + dx[i];
                int ty = cur.y + dy[i];
                
                if(tx<0 || tx>=N || ty<0 || ty>=M) continue;
                if(visited[tx][ty]) continue;
                if(maps[tx][ty] == 0) continue;
                
                visited[tx][ty] = true;
                q.offer(new Node(tx, ty, cur.cnt+1));
                
            }
        }
        
        return -1;
        
    }
}